packages/next-swc/crates/next-custom-transforms/src/transforms/page_static_info/collect_exports_visitor.rs

```diff
@@ -0,0 +1,183 @@
+use std::collections::HashSet;
+
+use lazy_static::lazy_static;
+use swc_core::ecma::{
+    ast::{
+        Decl, ExportDecl, ExportNamedSpecifier, ExportSpecifier, Expr, ExprOrSpread, ExprStmt, Lit,
+        ModuleExportName, ModuleItem, NamedExport, Pat, Stmt, Str, VarDeclarator,
+    },
+    visit::{Visit, VisitWith},
+};
+
+use super::{ExportInfo, ExportInfoWarning};
+
+lazy_static! {
+    static ref EXPORTS_SET: HashSet<&'static str> = HashSet::from([
+        "getStaticProps",
+        "getServerSideProps",
+        "generateImageMetadata",
+        "generateSitemaps",
+        "generateStaticParams",
+    ]);
+}
+
+pub(crate) struct CollectExportsVisitor {
+    pub export_info: Option<ExportInfo>,
+}
+
+impl CollectExportsVisitor {
+    pub fn new() -> Self {
+        Self {
+            export_info: Default::default(),
+        }
+    }
+}
+
+impl Visit for CollectExportsVisitor {
+    fn visit_module_items(&mut self, stmts: &[swc_core::ecma::ast::ModuleItem]) {
+        for stmt in stmts {
+            if let ModuleItem::Stmt(Stmt::Expr(ExprStmt {
+                expr: box Expr::Lit(Lit::Str(Str { value, .. })),
+                ..
+            })) = stmt
+            {
+                if value == "use server" {
+                    let export_info = self.export_info.get_or_insert(Default::default());
+                    export_info.directives.insert("server".to_string());
+                }
+                if value == "use client" {
+                    let export_info = self.export_info.get_or_insert(Default::default());
+                    export_info.directives.insert("client".to_string());
+                }
+            }
+
+            stmt.visit_children_with(self);
+        }
```

Comment:
That's not completely correct. e. g.

``` js
console.log("hello");
"use client"; // is NOT a directive
```

Reply:
```suggestion
        let mut stmts = stmts.iter();
        while let Some(stmt) in stmts.next() {
            stmt.visit_children_with(self);
            if let ModuleItem::Stmt(Stmt::Expr(ExprStmt {
                expr: box Expr::Lit(Lit::Str(Str { value, .. })),
                ..
            })) = stmt
            {
                if value == "use server" {
                    let export_info = self.export_info.get_or_insert(Default::default());
                    export_info.directives.insert("server".to_string());
                }
                if value == "use client" {
                    let export_info = self.export_info.get_or_insert(Default::default());
                    export_info.directives.insert("client".to_string());
                }
            } else {
                break;
            }
        }
        for stmt in stmts {
            stmt.visit_children_with(self);
        }
```

Reply:
https://github.com/swc-project/swc/pull/8617/files#diff-0b9e9189b20220866cd2f27040bfb128afdc5f94b71246da2604f598281a9b11 can be useful for this task